### PR TITLE
Fix createImportSpecifier with typescript 4.5+

### LINKED ISSUE #7325: [bug] Syntax error in TypeScript type defination files
### `package.json`

```json
{
    "source": "src/index.ts",
    "main": "dist/index.js",
    "module": "dist/index.esm.js",
    "types": "dist/index.d.ts",
    "dependencies": {
        "@swc/helpers": "^0.2.14",
        "web-utility": "^2.9.6"
    },
    "peerDependencies": {
        "katex": "^0.15.1",
        "react": ">=16",
        "react-bootstrap": "^1.6.4"
    },
    "devDependencies": {
        "@parcel/packager-ts": "^2.0.1",
        "@parcel/transformer-typescript-types": "^2.0.1",
        "@types/jest": "^27.0.2",
        "@types/katex": "^0.11.1",
        "@types/react": "^16.14.21",
        "husky": "^7.0.4",
        "jest": "^27.3.1",
        "katex": "^0.15.1",
        "lint-staged": "^12.0.3",
        "open-cli": "^7.0.1",
        "parcel": "^2.0.1",
        "prettier": "^2.4.1",
        "react": "^16.14.0",
        "react-bootstrap": "^1.6.4",
        "ts-jest": "^27.0.7",
        "typedoc": "^0.22.9",
        "typedoc-plugin-mdn-links": "^1.0.4",
        "typescript": "^4.5.2"
    },
    "browserslist": "> 0.5%, last 2 versions, not dead",
    "targets": {
        "main": {
            "optimize": false
        }
    }
}
```

### `tsconfig.json`

```json
{
    "compilerOptions": {
        "target": "ES5",
        "esModuleInterop": true,
        "jsx": "react",
        "lib": ["ES2021", "DOM"]
    },
    "include": ["src/**/*"]
}
```

## 🤔 Expected Behavior

Generate right **Type Defination** files for libraries, as what Parcel 2.0.0 did.

## 😯 Current Behavior

`dist/index.d.ts`

```typescript
import { PickData as  } from "web-utility";
import { PropsWithoutRef as , RefCallback as , PropsWithChildren as  } from "react";
import { type _PickData1 as  } from "web-utility/source/data";
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.1 |
| TypeScript | 4.5.2 |
| Node             | 14.17.6 |
| Yarn         | 1.22.15 |
| Operating System | Windows 10 Pro 21H1 |

### PR DESCRIPTION
**Pull request by astegmaier**
# ↪️ Pull Request

This fixes #7325 and possibly also #7400. The issue, as @benpigchu pointed out [here](https://github.com/parcel-bundler/parcel/issues/7325#issuecomment-988697578), was that Typescript 4.5 added support for [type modifiers on specific import names](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#type-modifiers-on-import-names). For example, it is now valid to write:
```ts
import { someFunc, type BaseType } from "./some-module.js";
```
Unfortunately, the way that the compiler API was changed to support this is not backward compatible :-(

Here's what the pre-4.5 API looked like:
```ts
createImportSpecifier(propertyName: Identifier | undefined, name: Identifier): ImportSpecifier;
```
And here's the new API ([see PR](https://github.com/microsoft/TypeScript/pull/45998)):
```ts
createImportSpecifier(isTypeOnly: boolean, propertyName: Identifier | undefined, name: Identifier): ImportSpecifier;
```
Since there's no way that I know of to check a function's arity at runtime, I had to resort to a version check.

There were several other backward-incompatible changes in that PR (to `updateImportSpecifier`, `createExportSpecifier` and `updateExportSpecifier`), and I verified that we don't use these APIs in parcel.

## 💻 Examples

Without this change, when you have typescript 4.5+, `@parcel/transformer-typescript-types` will generate output like:
```ts
import { something as } from "./foo"
```
instead of
```ts
import { something } from "./foo
```
This change ensures that we do the right thing.

## 🚨 Test instructions

Several of the existing unit tests will break if you upgrade to typescript 4.5+ and don't apply this change, so it didn't seem necessary to add new ones.

I manually verified that with this change, and typescript 4.2.4 (the old version in the repo), all the unit tests still pass. You can test it yourself by modifying `yarn.lock` to revert back to `4.2.4`, running `yarn install`, then `yarn test:integration`.

## ✔️ PR Todo

- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

### CODE REVIEW

File: packages/transformers/typescript-types/src/utils.js [hunk 1]
```diff
@@ -26,3 +27,22 @@ export function isDeclaration(ts: TypeScriptModule, node: any): boolean {
...
+  ts: TypeScriptModule,
+  propertyName: Identifier | void,
+  name: Identifier,
+  isTypeOnly: boolean = false,
+): ImportSpecifier {
+  const [majorVersion, minorVersion] = ts.versionMajorMinor
+    .split('.')
+    .map(num => parseInt(num, 10));
+  // The signature of createImportSpecifier had a breaking change in Typescript 4.5.
+  // see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#type-modifiers-on-import-names
```
Comment (devongovett) [inline packages/transformers/typescript-types/src/utils.js:41]: Wowwwwww... TS doesn't follow semver?? Maybe we should pin our dependency on it?

**Review by devongovett — APPROVED** [review]

File: packages/transformers/typescript-types/src/utils.js (hunk 1 above)
Comment (mischnic) [reply packages/transformers/typescript-types/src/utils.js:41]: The other side of that coin: https://github.com/parcel-bundler/parcel/issues/5978

**Comment by Jack-Works** [comment]
> there's no way that I know of to check a function's arity at runtime

Yes you can

image

**Comment by mischnic** [comment]
Though that would break if Typescript implemented `createImportSpecifier` using something like `createImportSpecifier(...args)` in the future.

**Pull request merged**